application/forms/ContactGroupForm.php

```diff
@@ -0,0 +1,284 @@
+            'submit',
+            [
+                'label' => $this->contactgroupId
+                    ? $this->translate('Save Changes') : $this->translate('Add Contact Group')
+            ]
+        );
+
+        if ($this->contactgroupId) {
+            $removeBtn = new SubmitElement(
+                'remove',
+                [
+                    'label'             => $this->translate('Remove'),
+                    'class'             => 'btn-remove',
+                    'formnovalidate'    => true
+                ]
+            );
+
+            $this->registerElement($removeBtn);
+            $this->getElement('submit')->prependWrapper((new HtmlDocument())->setHtmlContent($removeBtn));
+        }
+    }
+
+    /**
+     * Check if the cancel button has been pressed
+     *
+     * @return bool
+     */
+    public function hasBeenRemoved(): bool
+    {
+        $btn = $this->getPressedSubmitElement();
+        $csrf = $this->getElement('CSRFToken');
+
+        return $csrf !== null && $csrf->isValid() && $btn !== null && $btn->getName() === 'remove';
+    }
+
+    /**
+     * Get part updates
+     *
+     * @return array
+     */
+    public function getPartUpdates(): array
+    {
+        $this->ensureAssembled();
+
+        return $this->termInput->prepareMultipartUpdate($this->request);
+    }
+
+    /**
+     * Validate the terms
+     *
+     * @param Term[] $terms
+     *
+     * @return void
+     */
+    protected function validateTerms(array $terms): void
+    {
+        $contactTerms = [];
+        foreach ($terms as $term) {
+            $searchValue = $term->getSearchValue();
+            if (! is_numeric($searchValue)) {
+                $term->setMessage($this->translate('Is not a contact'));
+
+                continue;
+            }
+
+            $contactTerms[$searchValue] = $term;
+        }
+
+        if (! empty($contactTerms)) {
+            $contacts = (Contact::on(Database::get()))
+                ->filter(Filter::equal('id', array_keys($contactTerms)));
+
+            foreach ($contacts as $contact) {
+                $contactTerms[$contact->id]
+                    ->setLabel($contact->full_name)
+                    ->setClass('contact');
+            }
+        }
+    }
+
+    /**
+     * Load a contact group and populate the form
+     *
+     * @param int $groupId
+     *
+     * @return $this
+     */
+    public function loadContactgroup(int $groupId): self
+    {
+        $this->contactgroupId = $groupId;
+
+        $query = Contactgroup::on(Database::get())
+            ->columns(['id', 'name'])
+            ->filter(Filter::equal('id', $groupId));
+
+        $group = $query->first();
+        if ($group === null) {
+            $this->httpNotFound(t('Contact group not found'));
+        }
+
+        $groupMembers = [];
+        foreach ($group->contact->columns('id') as $contact) {
+            $groupMembers[] = $contact->id;
+        }
+
+        $this->storedValues = [
```

Comment:
Don't store this temporarily, as discussed. Load the stored data in the same transaction as when applying changes.